test_apps/smokeTests/gridTools.cpp

```diff
@@ -0,0 +1,439 @@
+    }
+
+    float *buf = new float[nblocks * block_size];
+
+    Heap.push_back(buf);
+
+    std::vector <float *> blks;
+    for (size_t i=0; i<nblocks; i++) {
+        blks.push_back(buf + i*block_size);
+    }
+
+    return(blks);
+}
+
+void MakeTriangle( Grid* grid, float minVal, float maxVal ) {
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+    
+    float value = minVal;
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                value = value==minVal ? maxVal : minVal;
+                grid->SetValueIJK( i, j, k, value );
+            }
+        }
+    }
+}
+
+void MakeConstantField( Grid* grid, float value ) {
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                grid->SetValueIJK( i, j, k, value );
+            }
+        }
+    }
+}
+
+void MakeRamp( Grid* grid, float minVal, float maxVal, size_t axis=X ) {
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+
+    float xIncrement = axis==X ? (maxVal - minVal) / (dims[X]-1) : 0;
+    float yIncrement = axis==Y ? (maxVal - minVal) / (dims[Y]-1) : 0;
+    float zIncrement = axis==Z ? (maxVal - minVal) / (dims[Z]-1) : 0;
+
+    float value = minVal;
+    for ( size_t k=0; k<z; k++ ) {
+        value = axis==Y ? minVal : value;      // reset value if we're ramping on Y
+        for ( size_t j=0; j<y; j++ ) {
+            value = axis==X ? minVal : value;  // reset value if we're ramping on X
+            for ( size_t i=0; i<x; i++ ) {
+                grid->SetValueIJK( i, j, k, value );
+                value += xIncrement;
+            }
+            value += yIncrement;
+        }
+        value += zIncrement;
+    }
+}
+
+void Foo(VAPoR::Grid* grid, double &rms, size_t &mv, size_t &d, bool nudge) {
+    cout << "foo " << rms << " " << mv << " " << d << endl;
+}
+
+void CompareIndexToCoords( 
+    VAPoR::Grid* grid, 
+    double &rms,                // Root Mean Square error
+    size_t &numMissingValues,   // Counter for receiving MissingValue upon query
+    size_t &disagreements,      // Counter for when AccessIJK() and GetValue() disagree
+    bool nudge
+) {
+    rms   = 0.f;
+    disagreements  = 0;
+    numMissingValues = 0;
+
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+ 
+    // Layered and Curvilinear grids have a bug where querying indices 
```

Comment:
What is the bug? How do you reproduce it?

Reply:
The bug is that Grid::GetValueAtIndex (and AccessIJK) reports wrong indices for Layered and Curvilinear grids.

I've made some changes to make testing this more clear.  After pulling and building the current code base, run the following commands

testGrid -arrangement Ramp

This will test coordinate vs IJK GetValue() calls for:
1) linear interpolation
2) nearest neighbor
3) nearest neighbor with nudging

Take a look at the results for the layered grid, which I've pasted below.  Nudging fixes the RMS error in this case.


Layered 8x8x8 Ramp up on Z axis:
  linear:
    RMS error:                                     0.1830
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  45

  nearestNeighbor:
    RMS error:                                     0.1603
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  90

  nearestNeighbor (nudged):
    RMS error:                                     0.0000
    Missing value count:                           0
    AccessIJK() vs GetValue() disagreement count:  0